### PR TITLE
xfstests: Support xfstests test on nfs

### PR DESCRIPTION
**Pull request by SeroSun**
Add support NFS test in xfstests, for now, Server and client test in a single VM.
Configure openqa parameter XFSTESTS=nfs to trigger tests under NFS filesystem. And 1 more parameter XFSTESTS_NFS_VERSION to set which NFS version to test, the default version is 4.1.

- Related ticket: https://progress.opensuse.org/issues/126464
- Verification run: 
1. xfstests for nfs
Run xfstests/nfs/001 in NFS4.1: http://10.67.133.133/tests/158
Run xfstests/generic/001-010 in NFS4.1: http://10.67.133.133/tests/157 (skip tests are known issue)

2. Verification run for inject code(in case some change in wrapper): http://10.67.133.133/tests/168 (inject a test code(free) in nfs/001:49)

3. Verification run for previous xfs/btrfs/ext4 tests
xfs/001-100: http://10.67.133.133/tests/164

### CODE REVIEW

File: data/xfstests/wrapper.sh [hunk 1]
```diff
@@ -2,14 +2,18 @@
...
 PROG="$XFSTESTS_DIR/check"
 SCRIPT_DIR=$(realpath $(dirname "$0"))
-if [ "$#" -gt 2 ]; then
-    INJECT_LINE=$2
-    INJECT_CODE=$3
+OPTIONS=""
+if [ "$#" -gt 1 ]; then
+    OPTIONS=$2
+fi
+if [ "$#" -gt 3 ]; then
```
Comment (lansuse) [inline data/xfstests/wrapper.sh:9]: This line should sync with line 87 of original file. But IMO use one sentence is more elegant.

File: data/xfstests/wrapper.sh [hunk 2]
```diff
@@ -90,7 +94,7 @@ fi
 
 # Run test
 log_file="/tmp/xfstests-$(echo "$1" | tr '/' '_').tmp"
-./$(basename "$PROG") -d "$1" | tee "$log_file"
+./$(basename "$PROG") -d "$OPTIONS" "$1" | tee "$log_file"
```
Comment (lansuse) [inline data/xfstests/wrapper.sh:97]: This will add one more space if $OPTIONS="". Although I know this doesn't matter

**Review by lansuse — COMMENTED** [review]
Its cause issue if use inject code

File: data/xfstests/wrapper.sh (hunk 2 above)
Comment (SeroSun) [reply data/xfstests/wrapper.sh:97]: Yes, I aware of this, a duplicate space will not harm to the original command.

File: data/xfstests/wrapper.sh (hunk 1 above)
Comment (SeroSun) [reply data/xfstests/wrapper.sh:9]: Added more verify scenarios. In the 87th line, it used $1 not affected by this options(which is $2), and the following parameters are all with para_names, so not affected too. But I do missing the parameter for inject_info, I add it in a new commit.

File: tests/xfstests/partition.pm [hunk 3]
```diff
@@ -281,6 +289,38 @@ sub format_with_options {
...
+        assert_script_run("sed -i 's/NFSV4LEASETIME=\"\"/NFSV4LEASETIME=\"$nfsgrace\"/' /etc/sysconfig/nfs");
+        assert_script_run("echo -e '[nfsd]\\ngrace-time=$nfsgrace\\nlease-time=$nfsgrace' > /etc/nfs.conf.local");
+    }
+    assert_script_run('exportfs -a && systemctl restart rpcbind && systemctl enable nfs-server.service && systemctl restart nfs-server');
+
+    # There's a graceful time we need to wait before using the NFS server
+    my $gracetime = script_output('cat /proc/fs/nfsd/nfsv4gracetime;');
+    sleep($gracetime * 2);
+}
+
```
Comment (pevik) [inline tests/xfstests/partition.pm:323]: The code is *almost* the same as the one in `tests/xfstests/partition.pm`. I wonder if it made sense to create `lib/xfstests.pm` and put function there.

**Review by pevik — APPROVED** [review]
Feel free to ignore my comments, it's good enough for merging. This is a good start, but it'd be great to add also multimachine test, which would run over network.

File: tests/xfstests/run.pm [hunk 4]
```diff
@@ -255,7 +255,10 @@ sub tests_from_ranges {
 sub test_run {
     my $test = shift;
     my ($category, $num) = split(/\//, $test);
-    my $cmd = "\n$TEST_WRAPPER '$test' | tee $LOG_DIR/$category/$num; ";
+    my $run_options = '';
+    $run_options = '-nfs' if check_var('XFSTESTS', 'nfs');
+    my $inject_code = get_var('INJECT_INFO', '');
+    my $cmd = "\n$TEST_WRAPPER '$test' $run_options $inject_code | tee $LOG_DIR/$category/$num; ";
```
Comment (lansuse) [inline tests/xfstests/run.pm:261]: Where is inject_line?

**Review by lansuse — COMMENTED** [review]
Lack of parameter of inject_line

File: tests/xfstests/partition.pm [hunk 5]
```diff
@@ -24,6 +24,7 @@ use File::Basename;
 
 my $INST_DIR = '/opt/xfstests';
 my $CONFIG_FILE = "$INST_DIR/local.config";
+my $NFS_VERSION = get_var('XFSTESTS_NFS_VERSION', '4.1');
```
[the remark this answers is not in the document]
Comment (SeroSun) [reply tests/xfstests/partition.pm:27]: > I'd always ask for a version (one will not have to search for the default), or print somewhere (in `setup_nfs_server()` ?) used version. Or do you plan put the NFS version into the name the testsuite?

+1, I will put the server version in testsuite name in o3 and osd.

File: tests/xfstests/partition.pm (hunk 3 above)
Comment (SeroSun) [reply tests/xfstests/partition.pm:323]: > The code is _almost_ the same as the one in `tests/xfstests/partition.pm`. I wonder if it made sense to create `lib/xfstests.pm` and put function there.

The code you look into is just what we use for xfstests, no worry about duplicate :D
But if further tests still could use this code, I'll consider reorganizing it to lib folder in that moment. Thank you for the suggestion!

**Comment by SeroSun** [comment]
> Feel free to ignore my comments, it's good enough for merging. This is a good start, but it'd be great to add also multimachine test, which would run over network.

+1, I have less experience in multimachine tests. I would suggest to put this as Version 1, and find suitable time to deploy another multimachine NFS test.

File: tests/xfstests/run.pm (hunk 4 above)
Comment (SeroSun) [reply tests/xfstests/run.pm:261]: They are in the same parameter, e.g. INJECT_INFO='42 do_something'

**Review by lansuse — APPROVED** [review]

**Review by czerw — APPROVED** [review]

**Review by schlad — APPROVED** [review]

**Pull request merged**